src/Features/CSharp/Portable/SignatureHelp/AbstractCSharpSignatureHelpProvider.LightweightOverloadResolution.cs

```diff
@@ -5,297 +5,252 @@
-                int position,
-                SemanticModel semanticModel,
-                ISemanticFactsService semanticFactsService)
-            {
-                if (arguments.Count != 0)
-                {
-                    foreach (var method in methodGroup)
-                    {
-                        if (FindParameterIndexIfCompatibleMethod(arguments, method, position, semanticModel, semanticFactsService, out var parameterIndex))
-                            return (method, parameterIndex);
-                    }
-                }
+            // Note: Providing no recommendation if no arguments allows the model to keep the last implicit choice
+            return (null, -1);
+        }
 
-                // Note: Providing no recommendation if no arguments allows the model to keep the last implicit choice
+        /// <summary>
+        /// Simulates overload resolution with the arguments provided so far and determines if you might be calling this overload.
+        /// Returns true if an overload is acceptable. In that case, we output the parameter that should be highlighted given the cursor's
+        /// position in the partial invocation.
+        /// </summary>
+        private (IMethodSymbol? method, int parameterIndex) TryFindParameterIndexIfCompatibleMethod(IMethodSymbol method)
+        {
+            // map the arguments to their corresponding parameters
+            using var argumentToParameterMap = TemporaryArray<int>.Empty;
+
+            for (var i = 0; i < arguments.Count; i++)
+                argumentToParameterMap.Add(-1);
+
+            if (!TryPrepareArgumentToParameterMap(method, ref argumentToParameterMap.AsRef()))
                 return (null, -1);
-            }
 
-            /// <summary>
-            /// Simulates overload resolution with the arguments provided so far and determines if you might be calling this overload.
-            /// Returns true if an overload is acceptable. In that case, we output the parameter that should be highlighted given the cursor's
-            /// position in the partial invocation.
-            /// </summary>
-            internal static bool FindParameterIndexIfCompatibleMethod(
-                SeparatedSyntaxList<ArgumentSyntax> arguments,
-                IMethodSymbol method,
-                int position,
-                SemanticModel semanticModel,
-                ISemanticFactsService semanticFactsService,
-                out int foundParameterIndex)
+            // verify that the arguments are compatible with their corresponding parameters
+            var parameters = method.Parameters;
+            for (var argumentIndex = 0; argumentIndex < arguments.Count; argumentIndex++)
             {
-                // map the arguments to their corresponding parameters
-                var argumentCount = arguments.Count;
-                using var _ = ArrayBuilder<int>.GetInstance(argumentCount, fillWithValue: -1, out var argToParamMap);
```

Comment:
a common problem was using fillWithValue, but then not ensuring the Count of hte array matches the count requested.  unless .Add is actually called the right number of times, that won't be the case.  I changed this to just get empty arrays and add hte number of fill-with-value values manually.